### PR TITLE
Copy of #2582, Added proxy parameters for ClientSession

### PR DESCRIPTION
**Pull request by RatanShreshtha**
Added proxy parameters for ClientSession so that subsequent requests can use the session proxy.

## Are there changes in behavior for the user?

New parameters of ClientSession():
- proxy
-  proxy_auth
-  proxy_headers

```
proxy_auth = aiohttp.BasicAuth('user', 'pass')
async with ClientSession(proxy="http://proxy.com", proxy_auth=proxy_auth) as session:
    async with session.get('http://test.com') as resp:
        print(await resp.text())
```
## Related issue number

#2580 
#2582 
## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [x] Documentation reflects the changes
- [x] If you provide code modification, please add yourself to `CONTRIBUTORS.txt`
  * The format is &lt;Name&gt; &lt;Surname&gt;.
  * Please keep alphabetical order, the file is sorted by names. 
- [x] Add a new news fragment into the `CHANGES` folder
  * name it `<issue_id>.<type>` for example (588.bugfix)
  * if you don't have an `issue_id` change it to the pr id after creating the pr
  * ensure type is one of the following:
    * `.feature`: Signifying a new feature.
    * `.bugfix`: Signifying a bug fix.
    * `.doc`: Signifying a documentation improvement.
    * `.removal`: Signifying a deprecation or removal of public API.
    * `.misc`: A ticket has been closed, but it is not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example: "Fix issue with non-ascii contents in doctest text files."

### CODE REVIEW

File: CONTRIBUTORS.txt [hunk 1]
```diff
@@ -240,4 +242,4 @@ Yuriy Shatrov
 Yury Selivanov
 Yusuke Tsutsumi
 Марк Коренберг
-Семён Марьясин
+Семён Марьясин
```
Comment (webknjaz) [inline CONTRIBUTORS.txt:272]: This change is unrelated. Fix your editor to preserve original EOF.

File: aiohttp/client.py [hunk 2]
```diff
@@ -281,6 +289,16 @@ def request(self,
         history = []
         version = self._version
 
+        # Merge with session proxy
+        proxy = self._proxy_info['proxy'] \
```
Comment (webknjaz) [inline aiohttp/client.py]: Please use brackets instead of escaping EOL

File: tests/test_proxy_session.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+from aiohttp import web
+
+
+@pytest.fixture
+def proxy_test_server(aiohttp_raw_server, loop, monkeypatch):
+    """Handle all proxy requests and imitate remote server response."""
+
+    _patch_ssl_transport(monkeypatch)
+
+    default_response = dict(
```
Comment (webknjaz) [inline tests/test_proxy_session.py]: Please use dict literal instead of a function

File: tests/test_proxy_session.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+        status=200,
+        headers=None,
+        body=None)
```
Comment (webknjaz) [inline tests/test_proxy_session.py]: Please use trailing comma and put the bracket on the next line

**Comment by asvetlov** [comment]
Not sure if I like yet another params in ClientSession constructor.

Anyway, the biggest showstopper now is proxy tests.
I have a strong feeling that mocking proxy calls has close to zero value.
Our functional proxy tests are messy too.
What we need to unblock any proxy improvement is implementing a test proxy server, rewriting tests to use it and dropping the current mocking approach.

**Comment by RatanShreshtha** [comment]
@asvetlov I am thinking about sending a separate PR with test proxy server and new tests.

**Comment by RatanShreshtha** [comment]
@webknjaz I am unable to solve the issue with mypy I have not used mypy before, can you guide me through that I have done wrong ?

**Comment by RatanShreshtha** [comment]
@webknjaz How can I correct miss matching types in code ?

Will type casting work ?

**Comment by RatanShreshtha** [comment]
@webknjaz @asvetlov Please review the PR.

File: docs/client_reference.rst [hunk 5]
```diff
@@ -48,7 +48,8 @@ The client session supports the context manager protocol for self closing.
                          timeout=sentinel, \
                          raise_for_status=False, \
                          connector_owner=True, \
-                         auto_decompress=True, proxies=None)
+                         auto_decompress=True, \
+                         proxy=None, proxy_auth=None, proxy_headers=None))
```
Comment (asvetlov) [inline docs/client_reference.rst]: Please document these params, not only enumerate in constructor signature.

File: aiohttp/client.py [hunk 6]
```diff
@@ -295,12 +303,22 @@ def request(self,
...
+        if proxy is not None and self._session_proxy is None:
             try:
                 proxy = URL(proxy)
             except ValueError:
                 raise InvalidURL(proxy)
 
+        if proxy is None and self._session_proxy is not None:
+            try:
+                proxy = self._session_proxy
+                proxy_auth = self._session_proxy_auth
```
Comment (asvetlov) [inline aiohttp/client.py:364]: By this line you are overriding explicitly passed `proxy_auth` and `proxy_headers` arguments with session default ones.

I have no idea how to resolve the conflict properly but proposed behavior looks confusing.

**Comment by webknjaz** [comment]
@RatanShreshtha tests fail now. This should be addressed.

**Comment by RatanShreshtha** [comment]
@webknjaz Please take a look I made the changes as suggested.

File: docs/client_reference.rst (hunk 5 above)
Comment (RatanShreshtha) [reply docs/client_reference.rst]: I have documented these changes.

File: aiohttp/client.py (hunk 6 above)
Comment (RatanShreshtha) [reply aiohttp/client.py:364]: I am trying to use session proxy if proxy is not mentioned explicitly for a particular request.

**Comment by webknjaz** [comment]
Hey @RatanShreshtha, this PR needs updating

**Comment by RatanShreshtha** [comment]
@webknjaz  let me update the PR

File: CONTRIBUTORS.txt [hunk 7]
```diff
@@ -91,6 +91,7 @@ Eugene Chernyshov
 Eugene Naydenov
 Eugene Tolmachev
 Evert Lammerts
+Fan Yu
```
Comment (webknjaz) [inline CONTRIBUTORS.txt:95]: Plz clean-up all unrelated changes.

File: CONTRIBUTORS.txt (hunk 1 above)
Comment (webknjaz) [reply CONTRIBUTORS.txt:272]: @RatanShreshtha ^

File: CONTRIBUTORS.txt (hunk 7 above)
Comment (RatanShreshtha) [reply CONTRIBUTORS.txt:95]: Should I squash all the changes ?

Comment (webknjaz) [reply CONTRIBUTORS.txt:95]: Up to you. We'll use "Merge and squash" strategy on this PR anyway :)

File: aiohttp/client.py [hunk 8]
```diff
@@ -344,12 +352,22 @@ def request(self,
             for i in skip_auto_headers:
                 skip_headers.add(istr(i))
 
-        if proxy is not None:
+        if proxy is not None and self._session_proxy is None:
```
Comment (asvetlov) [inline aiohttp/client.py:355]: What is the behavior if `proxy is not None and self._session_proxy is not None`?
I think passed `proxy` parameter should take precedence.
